---
title: "SP1 Analysis"
author: "Orfeas Gkourlias"
date: "3/10/2022"
output: pdf_document
---

```{r setup, results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(affy)
library(scales)
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(PoiClaClu)
library(edgeR)
library(EnhancedVolcano)
```

# 3. Exploratory Data Analysis

The data will be need to be extracted from the count file first. Since it's in .tsv format, the seperator is going to be tab based. The inclusion of headers adds an X to the sequence ID's, because R is unable to make headers out of just integers. To counteract this, the colnames will be manually added.

## 3.1 Loading the data
```{r}
file <- c("..\\data\\GSE152262_RNAseq_Raw_Counts.tsv")
# Raw_Data will be the primary dataframe that gets worked on.
raw_data <- read.table(file, sep = '\t', header = TRUE, row.names = 1)

# The first two and single last columns are the case samples.
# Control samples are indicated with con.
colnames(raw_data) <- c("case24275", "case24277", "con4279", "con4280", "con4280a", "case24281")

# Rearranging the columns so that the first three are the case samples.
raw_data <- raw_data[, c(1,2,6,3,4,5)]

# Showing the first five rows as an example.
raw_data[1:5,]

# Showing the dimension and structure of the raw_data dataframe.
dim(raw_data)
str(raw_data)
```
The data is now loaded in as a data frame. Every row shows the raw counts of a specific gene being expressed. 4275, 4277 and 4281 are the variant types. The datatypes are correct in this case_log2. There should only be integers included, except for the gene names.

Now that the data has been properly loaded, objects can be made to differentiate the control_log2 and case_log2 counts.
Before separating the groups, it'll be useful to apply a log2 function to our data. 
This makes it so that the data is more informative and tidier, because of outliers and the big range being worked with.

```{r}
# Transforming the read data of every columns to the log2 value
# 1 is added to every column to make sure there are no log2(0) values.
raw_data_log2 <- log2(raw_data + 1)

# Dividing the case and controls columns into seperate dataframes for later use.
case <- raw_data[,c(1:3)]
control <- raw_data[,c(3:6)]

# Applying the same division, but with the log values for plotting purposes.
case_log2 <- raw_data_log2[,c(1:3)]
control_log2 <- raw_data_log2[,c(4:6)]

# Displaying the first rows of divided dataframes.
case_log2[1,]
control_log2[1,]
```

The control_log2 and case_log2 data is now stored in different variables, as shown above.

## Visualizing using boxplot and density plot
More insight on the data can be gained by plotting and summarizing it.
Every column will first be summarized.
Following that, the mean values will be compared in a boxplot.

```{r}
# Applying a summary on all the log2 data.
summary(raw_data_log2)

# Getting the mean values of both controlled and case sample expression values
# For every gene. This might be useful later.
case_log2$mean = apply(X = case_log2[1:3], MARGIN = 1, FUN = mean)
control_log2$mean = apply(X = control_log2[1:3], MARGIN = 1, FUN = mean)

# Doing the same to the raw data frames.
case$mean = apply(X = case[1:3], MARGIN = 1, FUN = mean)
control$mean = apply(X = control[1:3], MARGIN = 1, FUN = mean)

# Plotting the log2 data.
boxplot(control_log2$mean, case_log2$m, outline = FALSE, names = c("control_log2 Mean", "case_log2 Mean"))
```
These boxplots are not yet very informative. 
The only thing that can be seen from them is that the case_log2s have a slightly lower expression level on average

Maybe a density plot allows for a more informative figure.
```{r}
# Creating the recurring colors that will be used for the columns
myColors <- hue_pal()(2)

# Density plotting the log2 data, using the colors created above.
plotDensity(raw_data_log2, col=rep(myColors, each=1), lty=c(1:ncol(raw_data_log2)),
            main = "Expression Distribution", xlab = "Log2(raw_data_log2 counts)")

# Adding a legend for clarity.
legend('topright', names(raw_data_log2), lty=c(1:ncol(raw_data_log2)),
       col=rep(myColors, each=3))
```
As can be seen in the plot, the highest amount of expressions, besides 0, seem to be around 10.

## 3.4 Visualizing using heatmap and MDS
Before continuing with this step, the data will have to be normalized.
There are 5 which rows are not actual genes. They will be removed.
After that, a barplot will be generated to show whether theres a difference in expression in millions, using col sums.

```{r}
# There are 5 rows which do not count for actual genes.
# These rows are currently not relevant, but shouldn't be present from now on.
remove_rows <- c("__not_aligned", "__no_feature", "__no_feature", 
                 "__alignment_not_unique", "__too_low_aQual", "__ambiguous")
raw_data <- raw_data[!(row.names(raw_data) %in% remove_rows),]

# Bar plotting the new data. Division by 1e6 shows the values in millions.
barplot(colSums(raw_data) / 1e6, las = 2, cex.names = 0.8, col = c(rep(myColors[1],3), rep(myColors[2],3)), xlab = "Genes", ylab = "Expression in millions")
```

Judging by that figure, the control group seems to have a higher average expression when summarised on all genes.

Now the DESeq2 library will be used to normalize the data. the VST function within this package is the next function.
The data will first have to make a Summarized Experiment object, which is done first.

```{r}
# Creating the dds Matrix, so that it ca nbe used in the vst function
(ddsMat <- DESeqDataSetFromMatrix(countData = raw_data,
                                  colData = data.frame(samples=names(raw_data)),
                                  design = ~ 1))

# Applying vst and saving it into the rld.dds object.
rld.dds <- vst(ddsMat)

# Applying assay on that object then saving it into rld.
rld <- assay(rld.dds)
```
Distance  calculation may now be performed on the normalized data.
The matrix will first have to be transposed.
After distance calculations have been performed, a heatmap may be constructed.
```{r}
# To create the heatmap, distances first get calculated & stored in a matrix
sampledists <- dist( t( rld ))
sampleDistMatrix <- as.matrix(sampledists)

# Annotation dataframe gets created for heatmap.
annotation <- data.frame(control = factor(rep(rep(1:3, each = 2), 1),
                                         labels = c("con4279", "con4280", "con4280a")))

# Rownames for the annotation get taken from raw_data.
rownames(annotation) <- names(raw_data)

# Heatmap function gets called on the matrix and annotation objects.
pheatmap(sampleDistMatrix, show_colnames = FALSE,
         annotation_col = annotation,
         clustering_distance_rows = sampledists,
         clustering_distance_cols = sampledists,
         main = "Euclidean Sample Distances")
```
The resulting heatmap shows where the large differences in expression are located.

The distances can also be shown using a 2d-plot, by performing multi dimensional scaling.
```{r}
# Creating the objects required by ggplot for mds.
dds <- assay(ddsMat)
poisd <- PoissonDistance( t(dds), type = "deseq")
samplePoisDistMatrix <- as.matrix(poisd$dd)
mdsPoisData <- data.frame( cmdscale(samplePoisDistMatrix) )

# Creating names for the coords.
names(mdsPoisData) <- c('x_coord', 'y_coord')

# Getting the columns as factors.
groups <- factor(rep(1:6, each=1), 
                 labels = names(raw_data))

# Column name extraction.
coldata <- names(raw_data)

# Plotting the distance data in a 2d plot with ggplot.
ggplot(mdsPoisData, aes(x_coord, y_coord, color = groups, label = coldata)) + 
  geom_text(size = 4) +
  ggtitle('Multi Dimensional Scaling') +
  labs(x = "Poisson Distance", y = "Poisson Distance") +
  theme_bw()
```
## 3.5 Cleaning Data
After examination of the case and control groups, there shouldn't be any samples removed.
This would also not be possible, because at least 3 samples are required per group.

# 4 Discovering Differentialy Expressed Genes (DEGs)    
Proceeding all the insight gained from plotting the data, it may now all be analysed in R. The purpose being is the discovery of DEGs, differentialy expressed genes. The earlier plots showed that there will most likely be plenty of those. The observed mutation also causes a frame shift, increasing the likelihood of DEGs greatly.. Before performing the analysis steps, the data will need to go through a pre-processing phase.

## 4.1 Pre-processing
First, the FPM, fragments per million mapped fragments, will be calculated for every row/gene.

```{r}
# Applying the FPM calculation then creating a data frame out of it.
raw_data.fpm <- log2( (raw_data/ (colSums(raw_data) / 1e6 )) + 1)
```

There are quite a lot of inactive genes within the dataset.
Filtering these out will help in further analysis.
The paper does not provide a method for filtering out these genes.
First, the most fitting method will have to be chosen.

Let's first see what the actual sum values are of all the genes.

```{r}
# Make a column which sums up the log2 reads.
raw_data.fpm$sum = apply(X = raw_data.fpm, MARGIN = 1, FUN = sum)

# Calculating the percentage of genes with a total of 0 counts across all groups
sum(raw_data.fpm$sum == 0) / nrow(raw_data.fpm) * 100
```
The calculation above returns a value of 40.55.
So 40.55% of the genes have not been expressed in any group or sample.
Before applying other calculations to detect more inactive genes, it can be concluded that this 40% is surely inactive.
It's therefore safe to remove.

There's not a definitive answer as to when a gene may be considered inactive and irrelevant.
Discussion is still ongoing, but an answer that was observed multiple times is that an FPM sum above 0.5 indicates a statistically considerable gene.
Let's see how much of the data is retained if everything below 0.5 would be removed.
```{r}
sum(raw_data.fpm$sum > 0.5)
sum(raw_data.fpm$sum > 0.5) / nrow(raw_data.fpm) * 100
```
This would result in 21586 genes remaining for further analysis. Which is 37.02% of the original data.
While the percentage is a little low, it's still 21 thousand genes, which is enough for analysis.
```{r}
raw_data.fpm <- raw_data.fpm[raw_data.fpm$sum > 0.5,]
```

# 4.2 The Fold Change Value
To gain insight into how the control and experiment groups differ in expression, a FC value will be calculated. First, to calculate the FC values, the averages will be calculated. These will be subtracted from eachother to get a LFC value. The FC value but with log2 applied.
```{r}
# Calculating the means and savin them in columns
raw_data.fpm$case_avg = apply(X = raw_data.fpm[,1:3], MARGIN = 1, FUN = mean)
raw_data.fpm$con_avg = apply(X = raw_data.fpm[,4:6], MARGIN = 1, FUN = mean)

# Calculating the LFC values.
raw_data.fpm$lfc = raw_data.fpm$case_avg - raw_data.fpm$con_avg
```

Following this, a histogram may be made out of the new values.
```{r}
hist(raw_data.fpm$lfc, breaks = 60)
abline(v = -1:1)
```
As can be seen in the histogram, there are quite some LFC values higher than 1, indicating that there is indeed increased expression on multiple genes. 

## 4.3 Using Bioconductor Packages
It's now possible to perform T tests to determine which genes are significantly different in expression. Instead of doing this manually, bioconductor packages can be utilized. Because this experiment used edgeR, it will also be used here.

The raw_data data frame will need to be used again, since edgeR requires the raw counts. Before trying to detect DEGs, some filtering can be applied to make the results more significant. As discussed in 4.1, filtering out low counts tends to be beneficial. edgeR has a built in function which uses it's own filtering algorithm to remove these low counts. This will be done after the dataframe has been converted to the appropriate format, the DGElist

```{r}
# Defining the DGElist object.Group indicates which rows are cases/control.
dge <- DGEList(counts = raw_data, group = c(2,2,2,1,1,1))
# Showing the amount of rows read from the raw counts (All rows).
nrow(dge)
# Marking the rows that should be kept because they have sufficient expression levels
keep <- filterByExpr(dge)
# Only keeping the rows which have sufficient expression. Showing the amount of rows kept.
dge <- dge[keep, , keep.lib.sizes=FALSE]
nrow(dge)
```
Now that the DGE object has been created and low counts have been filtered out, further normalization can be applied. In the case of this experiment, that wont be needed.

The classic edgeR pipeline will be followed here, since there's nothing that needs to be done to the count data after the filtering done above. The first step of the pipeline is to calculate gene dispersions. This can be done using the estimtaeDisp function.

```{r}
dge <- calcNormFactors(dge)
dge <- estimateDisp(dge)
dge
```
Now that the dispersions have been calculated, testing for the DE genes may be performed. This is done by using the ExactTest function, which looks at the two groups and performs a t.test, to then determine the P value. The genes which get assigned a P value of less than 0.01 shall be stored in a results dataframe. 0.01 Was used in the paper. In order of lowest to highest P value. The are the adjusted P values.

```{r}
et <- exactTest(dge, pair=c(1,2))
res <- topTags(et, n = Inf, p = 0.01)$table
```

Now the significant DEGs have been selected and assigned with LogFC values.
Further analysis can be done by plotting the results.

# 5 Data Analysis and Visualization
There's multiple ways to display the DEGs. Since the data set consist of gene counts for every gene, there'll be too many rows to properly display for some visualizations.
It might be beneficial to therefore look at the pathways to which these important genes belong to. But, a volcano plot can still be made out of the initial results.

## 5.1 Volcano Plot
A volcano plot can show which DEGs are of most importance with the given LogFC values. 
The paper also contains a volcano plot, which has a FC cutoff of 2. This cutoff point will also be used here.


```{r}
sum(res$logFC < 0)

deseq.volcano <- function(res, datasetName) {
  return(EnhancedVolcano(res, x = 'logFC', y = 'FDR',
                         lab=rownames(res),
                         title = paste(datasetName, "trisomic vs disomic"),
                         subtitle = bquote(italic('FDR <= 0.05 and absolute FC >= 2')),
                         # Change text and icon sizes
                         labSize = 3, pointSize = 1.5, axisLabSize=10, titleLabSize=12,
                         subtitleLabSize=8, captionLabSize=10,
                         # Disable legend
                         legendPosition = "none",
                         # Set cutoffs
                         pCutoff = 0.01, FCcutoff = 2))
}

deseq.volcano(res = res, datasetName = "res")

```





